File: Leetcode/1155. Number of Dice Rolls With Target Sum/1155. Number of Dice Rolls With Target Sum.py
class Solution:
    def numRollsToTarget(self, n: int, k: int, target: int) -> int:
        dicti={}
        m=(10**9)+7
        def rec(i,res):
            if i in dicti:
                if res in dicti[i]:
                    return dicti[i][res]
            if res>=target or res+n-i>target or res+(n-i)*k<target:
                return 0
            if i==n-1 and res+k>=target:
                return 1
            temp=0
            for j in range(1,k+1):
                temp+=rec(i+1,res+j)
            if i not in dicti:
                dicti[i]={}
            dicti[i][res]=temp
            return temp%m
        return rec(0,0)

File: Leetcode/1155. Number of Dice Rolls With Target Sum/1155. Number of Dice Rolls With Target Sum.java
class Solution {
    static int MOD = (int)1e9 + 7;
    int helper(int n,int t,int k,int dp[][]){
        if(n < 0 || t < 0) return 0;
        if(n == 0 && t == 0) return 1;
        if(dp[n][t] != -1) return dp[n][t];
        int c_ans = 0;
        for(int i = 1;i <= k;i++){
            c_ans = (c_ans + helper(n - 1,t - i,k,dp)) % MOD;
        }
        return dp[n][t] = c_ans;
    }
    public int numRollsToTarget(int n, int k, int target) {
        int dp[][] =  new int[n + 1][target + 1];
        for(var a:dp) Arrays.fill(a,-1);
        return helper(n,target,k,dp);
    }
}